### DISCUSSION TITLE
0056 XLS-56d: Atomic/Batch Transactions

### DISCUSSION THREAD

**Discussion by mvadari** [XRPLF/XRPL-Standards#162] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by dangell7** [comment]
I'm not sure I like the name `Atomic` if there is an option for it to be a non atomic tx.

I would also just want non atomic batch. Try all 8 and return the responses of each. I will parse the results and handle accordingly. 

If atomic proves to be very difficult this will be a much easier implementation

**Reply by mvadari** [reply]
The transactions are still being applied atomically, regardless of the scenario. But I'm certainly open to other names. I do think `Batch` is too specific for this transaction's capabilities, though.

I don't think it should be any more difficult to handle atomic-ness compared to just batching transactions - the batching part in general is where the complexity is.

**Comment by ajkagy** [comment]
my thoughts here

1. would it be possible to have a flag to submit all transactions regardless of a failure. I can see a use for this in terms of bulk buying nfts where as a user wants to create a buy offer for 20 nfts, but doesn't want 1 failure to stop the rest of batched buy offers from being submitted because 1 of the nfts changed ownership between the time of the user creating the offer and submitting it (edge case failure, but have seen it happen when multiple parties are competing to buy at the same time).
2. Fee calculation best practices? would this affect open ledger fee if say 100 or 1000 or more transactions are being submitted, therefore a buffer amount would need to be added for reliable transaction submission into open ledgers?
3. Since this is technically only 1 transaction, I would assume there's no submission node queue limit issue here?
4. What would be the batch size/payload limit?

**Reply by mvadari** [reply]
1. Yes, that could be another atomicity type.
2. That's the main reason this is limited to 8 transactions max, to avoid someone trying to batch 100 transactions and bloating ledgers. I think it would require more investigation, though.
3. I believe not, since this is only one transaction.
4. 8 transactions.

**Reply by dangell7** [reply]
Fee escalation is a great point! Would prob have to try, but fail imo. We could add a flag to charge the full fees though.

**Reply by ajkagy** [reply]
yea, fee escalation would be my concern since you can potentially stuff hundreds of transactions into a single open ledger (if the fee amount is sufficient) thereby compounding the number of transactions...more than likely fee amount will change and those transactions would be queued resulting in the user receiving a telCAN_NOT_QUEUE.

**Reply by mvadari** [reply]
Yeah, that's why I limited it to 8 transactions (mostly a randomly-picked number). But it definitely will require further testing.

**Comment by shortthefomo** [comment]
Am i reading it correctly that this actually reduces the fee for the most complex types on the ledger, large count multisig and escrow finish? 

Batch fee either needs to be dynamic based on tx types in the batch, or worst case assumption for all 8 slots.

**Reply by mvadari** [reply]
That's incorrect.
 $$2*base\textunderscore fee + \sum_{txs}tx.Fee$$
 This means that the fee for a batch transaction is 2 times the base fee (20 drops total) plus the sum of whatever the normal fee would be for the sub-transactions. So the cost of multisig and escrow finish would be incorporated into that.

**Comment by tequdev** [comment]
What will the value of `meta.deliverd_amount` when multiple Payments are sent?

There could be a impact on the exchange's deposit system.

But it is a very valuable feature for the ecosystem as a whole!

**Reply by mvadari** [reply]
That's a good question and will likely require some further thought and experimentation. Perhaps there should be a new field called `meta.delivered_amounts` for this purpose that has a list. Or a list of `metadata`s for the transaction as a whole.

**Comment by Tapanito** [comment]
I'd like to clarify the way transactions are counted. The spec mentions that it will be possible to nest the transactions, but the total number of transactions cannot exceed 8.
Imagine a user submits the following atomic transaction:
 - atomic(tx1,atomic(tx2, tx3))
 
 In the total transaction count is this considered 3 or 4 transactions? I.e. does the nested atomic contribute to the total transaction count?
 
 Also, is there any rationale behind allowing a user to submit 8 transactions as opposed to 10 or 6?

**Reply by dangell7** [reply]
8 is chosen arbitrary. I don't think I will support nested txns.

**Reply by Tapanito** [reply]
Could you elaborate on your position?

**Reply by dangell7** [reply]
Nesting batch transactions requires another level of complexity. I don't want to support this at this time. If you want to write this amendment with nested transactions feel free to PR the amendment. I would like to get this amendment finished ASAP and to do so will need to strip it down to something much simpler.

**Reply by Tapanito** [reply]
To address your commend regarding getting this done ASAP I'd like to say the following. This feature introduces fairly complex interactions with the rest of XRP Ledger. While I appreciate the enthusiasm to implement  the spec before it is finalized, I don't think a feature specification should follow development, but the development should follow the specification. Otherwise, what is the purpose of writing a specification in the first place.  I'm sure you're very aware that any future changes to this feature would have to go through the same long amendment process. After all, once a feature is "in", it's very very difficult to remove in case of any oversight.

**Reply by mvadari** [reply]
The spec has been updated to remove nesting for now, since most of the benefit of this feature does not require nesting, and nesting becomes rather complicated from an implementation perspective. It can be re-introduced in the future, based on need.

**Comment by Tapanito** [comment]
This feature supports nesting, which means we can do some arbitrary conditional transaction execution by combining different atomic types. IHMO this is pretty cool, and versatile. Who needs smart contracts when you have logic, am I right? :P 

 We already have an ${x \land y}$ from **ALL**, ${x \lor y}$  from **BATCH** all is missing is ${\neg x}$. Then for completeness I suggest to introduce **NONE** atomic type, which is successful only when all transactions fail. On it's own such atomic operation is useless. However, since the atomic transactions support nesting, and there already is an **and** in the form of **ALL** and **or** in the form of **BATCH** then introducing **NONE** would allow to the users to do rudimentary logical operations with transactions.

For example, with 8 transactions users would already be able to express secondary boolean operations such as:

${\textstyle x\rightarrow y=\neg {x}\vee y}$
${\textstyle x\leftrightarrow y=(x\land y)\lor (\neg x\land \neg y)=(x\lor \neg y)\land (\neg x\lor y)}$
${\textstyle x\oplus y=\neg (x\leftrightarrow y)=(x\vee y)\wedge \neg (x\wedge y)=(x\vee y)\wedge (\neg x\vee \neg y)=(x\wedge \neg y)\vee (\neg x\wedge y)}$

**Reply by mvadari** [reply]
The spec has been updated to remove nesting for now, as this would add a lot of implementation complexity, and the network still gets a lot of benefit from unnestable `Atomic` transactions. This additional atomicity type can be added alongside nesting.

**Comment by tequdev** [comment]
For example, when using OfferCreate, Sequence is a component of OfferID. If multiple OfferCreate transactions are used in this transaction, what will the OfferID?

**Comment by tequdev** [comment]
If AtmicType is ONLYONE or BATCH, it might be helpful to see which transactions were successful (meta field).

**Comment by mvadari** [comment]
The spec was significantly updated on 2024-04-22.

**Comment by nhartner** [comment]
I presume ticket sequences are intentionally excluded, which I think is good to reduce scope if they can still be supported in a future version. I could see Tickets being useful for atomic swaps because until all parties sign and submit, their accounts cannot submit any other transactions to the ledger otherwise they will invalidate their sequence numbers inside the batch transaction.

**Reply by mvadari** [reply]
Inner transactions do not need sequence numbers at all, to avoid this exact problem. See section 2.3 and section 4.

**Reply by nhartner** [reply]
I'm referring to the `Sequence` field of `BatchTxn`. For the multiple account scenario, I understood that to be the sequence number for each account. So if Bob assigns sequence 111 into his inner transaction BatchTxn, and Alice and assigns sequence 222, then neither Bob nor Alice can submit any transactions on their account until both have signed the Batch transaction and it has been submitted to the ledger.

**Reply by mvadari** [reply]
Ticket support for that field would be rather complicated, given the way that the sequence numbers are used (see section 4.2).

**Comment by nhartner** [comment]
```the RawTransactions field is not included in the validated version of the outer transaction.```
If the RawTransactions field is not included in the validated ledger, doesn't it become impossible to verify the signature of a validated batch transaction that used ONLYONE or UNTILFAILURE? Since skipped transactions from the batch will not be in the validated ledger but were part of the signature?

**Reply by mvadari** [reply]
From section 2.3:

> This field is not included in the validated transaction, **nor is it used to compute the outer transaction signature(s)**

It is not included in the signature either.

**Reply by nhartner** [reply]
Ah got it. So the TxIDs field of the validated ledger would have the hashes of all the raw transactions, even the ones that didn't get processed.

**Reply by mvadari** [reply]
Yes, exactly.

**Comment by rothexD** [comment]
regarding FAQ 3, would this not guarantee that all transactions are part of the same ledger?

**Reply by mvadari** [reply]
Oops, I'll take that out - it's equivalent to the `INDEPENDENT` mode.

**Reply by rothexD** [reply]
true lol, missed that

**Comment by tequdev** [comment]
Although not directly related to the content, Outer Transaction Fee in 6.1.2 Sample Ledger is set to "0". Perhaps it should be "40".

**Reply by mvadari** [reply]
Good catch, fixed 🙂

**Comment by tequdev** [comment]
If Multiple Account Batch and Inner Txn contains multiple transactions for the same account, will BatchTxn.Sequence be incremented for each?

If so, Ledger Object ID should use `BatchTxn.Sequence` instead of `BatchTxn.Sequence + BatchTxn.BatchIndex`.

Even if it is not increased and is processed with the same value, `BatchTxn.BatchIndex` should not be used considering the case where Inner Txn is [Alice Txn, Bob Txn, Alice Txn]. There is a possibility of ID conflict in Alice's transaction immediately after Batch.

```
Account Sequence before Batch txn: 10
Ledger Object ID generated from Batch[0]: 10 (10+0)
Ledger Object ID generated from Batch[2]: 12 (10+2)
Account Sequence after Batch txn: 12

Ledger Object ID generated from some Txn: 12
```

**Reply by mvadari** [reply]
`BatchTxn.Sequence` will not be incremented for each, as that makes sequence numbers harder to deal with in inner transactions, since you don't always know how many transactions will be successful.

The batch transaction increments the sequence number by $numInnerTxns$ instead of just by 1 (see section 4.2). So the `Account Sequence after Batch txn` will actually be 13, not 12.

**Reply by tequdev** [reply]
Does this mean that Sequence: 11 will skipped on Alice's account in this case?

**Reply by mvadari** [reply]
Yes.

I know it's not the most elegant solution, but it's the best I could come up with, given that the sequence number has to be calculable just given info in the inner transactions. I'm open to other suggestions as well.

**Comment by shefi** [comment]
Is there any chance the processing could reach a deadlock state if there is any dependence between inner transactions or between inner and outer transaction?

I have no knowledge of how the transactions are being processed within rippled, so this might be a nothing burger.

I am just asking, because batch processing is a kind of traditional DB transactions and they are prone to deadlocks if not implemented right.

**Reply by mvadari** [reply]
All the dependencies are addressed in this spec - just fees and sequence numbers. That's all the outer transaction needs to do, really.

**Comment by EnchStyle** [comment]
Question. How does it work with wallet providers?
If XPMarket sets a UI fees of x%, and then pushes transaction to Xaman wallet who would like to also charge y% free.... How will it work? Will it be x+y%? just x%? just y%?

**Reply by mvadari** [reply]
You're limited to what's possible with existing transactions. There is no percentage transaction right now, so both the platform and the wallet would have to calculate that themselves based on the transaction they're handed.

**Comment by shortthefomo** [comment]
After spending a bit of time looking at XLS-67d. I'm in favor of splitting the charge out of the batch, and removing the documentation to use it for service charges in the batch.

In favor of both amendments moving forward with the caveat that charges use the chareback field vs bundling those into a batch.

Motivation is mainly interfaces having a clear place to derive / find and display clearly display additional charges. This could become complicated if batch was used obscurely.

**Reply by mvadari** [reply]
How would you prefer handling multiple parties charging on the same transaction (e.g. a platform and a wallet)? Seems easier to do that with Batch than with XLS-67d.

**Comment by VadneyK** [comment]
would love to see this go through. Trying to implement CRU (carbon credits) and CET (emissions data) linking and need atomic transactions.

CRU and CET are defined by GBBC IWA TTF

**Comment by mvadari** [comment]
Per the new [XLS Contributing process](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md), it is my opinion that we have reached a "well-refined standard."

As such, I propose that we move this discussion to a file (via https://github.com/XRPLF/XRPL-Standards/pull/197) and work on final changes using additional PRs, for better change-tracking.

Please comment here if you would like to object to moving this spec/discussion forward in the process into a DRAFT spec.

(Note that per the Contributing guidelines, moving a spec into the [DRAFT](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md#drafts) state does not mean any kind of endorsement, nor does it mean that this specification will become adopted. It is solely meant as a mechanism to enable better change tracking using PRs.)

**Comment by ckeshava** [comment]
1. What is the reason for picking `(n+2) * BaseFee` in the first half of the `Batch` transaction fees? I understand the multi-signature transactions have a factor of `(n+1)` in its fees. Why didn't you retain `n` or `n+1` multiplicative factor in the fee calculation?

2. What's the influence of `Batch` transactions on the open-ledger cost? For instance, if 10 `Batch` transactions are included in a validated ledger, this is equivalent to inserting at most 80 non-`Batch` "usual" transactions. Is the soft-limit of the open-ledger cost calculation proportionately increased by this amendment?

3. Regarding Section 4.2, the new Sequence numbers appears to *waste* a few of the intermediate sequence numbers. I know that this is too late to change the spec, but how about this alternative.

Instead of incrementing the *new_seq_num* by `numInnerTxns`, why can't we use: 
`new_seq_num_Acct1 = old_seq_num_Acct1 + numInnerTxn_Acct1`; `numInnerTxn_Acct1` refers to the number of inner-transactions sent by Account-1 only.

This does not completely eliminate wastage of the sequence numbers, however it reduces it. But, it comes with an increased computational overhead.

**Reply by mvadari** [reply]
The latest version of the spec is here: https://github.com/XRPLF/XRPL-Standards/pull/197

**Reply by ckeshava** [reply]
ok

**Reply by ckeshava** [reply]
what do you think of question-2 ?

**Comment by ckeshava** [comment]
5. I have some questions about the intended use-cases of this amendment. Can you point me to additional use-cases or test-cases that clarify the usage?

> Some use-cases that may be enabled by batch transactions include (but are certainly not limited to):
> All or nothing: Mint an NFT and create an offer for it in one transaction. If the offer creation fails, the NFT mint is reverted as well.

If the NFTokenOffer creation fails, I would retry the transaction after fixing the errors.

> Trying out a few offers: Submit multiple offers with different amounts of slippage, but only one will succeed.

If I submit multiple offers on the DEX (with varying slippage), the offer-matching algorithm ensures that both buyer and seller obtain the optimal match. I can submit multiple offers with the `tfPassive` flag. How is the `Batch` amendment helpful in this case?

> Withdrawing accounts (multi-account): Attempt a withdrawal from your checking account, and if that fails, withdraw from your savings account instead.

I can encapsulate the transaction submission code inside a try-catch or an if-else block. This helps me implement the required logic. Why would I need the `Batch` amendment?

6. Canonical order of transactions: This depends on the AccountID, Sequence number of the transaction and the transaction-ID Here's the [Source Code](https://github.com/XRPLF/rippled/blob/8429dd67e60ba360da591bfa905b58a35638fda1/src/ripple/app/misc/CanonicalTXSet.cpp#L25-L36). Why are `Batch` transactions more prone to front-running? I'm assuming that the signatories do not leak information themselves. Given that fact, the consensus process for `Batch` transactions does not appear to take up significantly higher time, compared to non-`Batch` transactions.

**Reply by mvadari** [reply]
> 5. I have some questions about the intended use-cases of this amendment. Can you point me to additional use-cases or test-cases that clarify the usage?

* Platform fees - e.g. a DEX platform can charge a platform fee alongside the `OfferCreate` transaction, and the user can't reject the platform fee because it's not a separate transaction.
* https://developer.algorand.org/docs/get-details/atomic_transfers/#use-cases
* https://x.com/msvadari/status/1770478996741874156
* https://x.com/LoafPickle/status/1770509374458364275

> > Some use-cases that may be enabled by batch transactions include (but are certainly not limited to):
> 
> > Trying out a few offers: Submit multiple offers with different amounts of slippage, but only one will succeed.
> 
> If I submit multiple offers on the DEX (with varying slippage), the offer-matching algorithm ensures that both buyer and seller obtain the optimal match. I can submit multiple offers with the `tfPassive` flag. How is the `Batch` amendment helpful in this case?

That doesn't help if you e.g. have 100 XRP and only want to spend 10 XRP. If you place multiple offers, they might all be matched.

> > Withdrawing accounts (multi-account): Attempt a withdrawal from your checking account, and if that fails, withdraw from your savings account instead.
> 
> I can encapsulate the transaction submission code inside a try-catch or an if-else block. This helps me implement the required logic. Why would I need the `Batch` amendment?

It's more efficient to essentially have that try-catch/if-else on-chain.

> 6. Canonical order of transactions: This depends on the AccountID, Sequence number of the transaction and the transaction-ID Here's the [Source Code](https://github.com/XRPLF/rippled/blob/8429dd67e60ba360da591bfa905b58a35638fda1/src/ripple/app/misc/CanonicalTXSet.cpp#L25-L36). Why are `Batch` transactions more prone to front-running? I'm assuming that the signatories do not leak information themselves. Given that fact, the consensus process for `Batch` transactions does not appear to take up significantly higher time, compared to non-`Batch` transactions.

They enable things like circular payments.

**Reply by ckeshava** [reply]
thanks

**Comment by ckeshava** [comment]
What is the maximum number of Signers allowed in a `Batch` transaction? Assuming that `featureExpandedSignerList` is enabled, what’s the upper bound on the signers ?

**Reply by ckeshava** [reply]
Based on a cursory read of this method, every inner transaction could have 32 unique signers: https://github.com/XRPLF/rippled/blob/develop/include/xrpl/protocol/STTx.h#L61

Is that correct ?

**Reply by ckeshava** [reply]
However, the `preflight` function restricts the number of unique signers to `8`

**Reply by mvadari** [reply]
There are two separate types of signers here - `BatchSigners` and `Signers`.

Each `BatchSigners` is for each account submitting an inner transaction. There can be a max of 8 of those (due to a max of 8 inner transactions).

Each `BatchSigners` may contain `Signers`, if that account has a multisign setup. There may be up to 32 `Signers` in each `BatchSigners`.

**Comment by ckeshava** [comment]
Few more questions:
1. How is transaction-retry mechanism handled with the Batch amendment? Suppose one of the inner-transactions result in a `ter` error code, will that transaction be retried as a "non-batch normal" transaction in the next ledger? 

2. From a user experience point-of-view, it is cumbersome to compute the transaction-IDs for all the `RawTransactions`. I understand that tooling libraries have the `sign` method for this purpose. But in a future revision, perhaps rippled can automatically compute the transaction-IDs from the RawTransaction array? Are there any problems with this design?

3. In a validated ledger, will the inner-transactions appear in one-block (along with the outer-batch transaction)? (or) are all the transactions displayed in the canonical order irrespective of their inner-batch/non-batch status? The examples in the spec only consider one batch-transaction in a ledger. I could not understand the final order of transactions in the validated ledger.

**Reply by ckeshava** [reply]
@legleux please share your thoughts here

**Reply by mvadari** [reply]
1. No, because the transactions depend on each other. All non-`tesSUCCESS` errors are treated as failures.
2. There was an independent discussion about this earlier today. This change will be made, and the transaction IDs will be added as a synthetic field to the RPCs.
3. The spec is outdated on this. The outer transaction has to happen before the inner transactions.

**Comment by mvadari** [comment]
Closing this discussion since the spec has been initially merged via https://github.com/XRPLF/XRPL-Standards/pull/197. Thank you to everybody who participated in this discussion.

For further discussion or changes, please open a PR in this repo on XLS-56d.
